### PR TITLE
New SPM product: UBQRScanner

### PR DESCRIPTION
**Pull request by ubfelix**
This PR adds a new product `UBQRScanner` to the UBKit library. The core of this product is a class `QRScannerView`, a `UIView` subclass that provides functionality related to the scanning of QR codes and other supported formats. Internally, it uses an `AVCaptureSession`.

Successfully scanned codes (as Strings) and errors are received via the `QRScannerViewDelegate` methods.

### CODE REVIEW

File: Sources/UBQRScanner/QRScannerView.swift [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+import AVFoundation
+import Foundation
+import UIKit
+
+/// A view that provides functionalty related to the scanning of QR codes and other supported formats,
+/// using the device's video camera. When started, the view displays the video camera feed. Events, like
+/// the successful scanning of a code or specific errors are received via the `QRScannerViewDelegate` methods.
+/// - Important: Apps using this view must provide a value for `NSCameraUsageDescription` in their `Info.plist`,
+/// else the app will crash as soon as the `startScanning()` method is called.
+public class QRScannerView: UIView {
```
Comment (joseph-elmallah) [inline Sources/UBQRScanner/QRScannerView.swift:17]: Why not open?

File: Sources/UBQRScanner/QRScannerView.swift [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+        NotificationCenter.default.addObserver(forName: UIApplication.willResignActiveNotification, object: nil, queue: nil) { [weak self] _ in
+            guard let self = self else { return }
+            self.lastIsRunning = self.isRunning
+            self.lastIsTorchOn = self.isTorchOn
+            self.stopScanning()
+        }
+        NotificationCenter.default.addObserver(forName: UIApplication.didBecomeActiveNotification, object: nil, queue: nil) { [weak self] _ in
+            guard let self = self else { return }
+            if let lastIsRunning = self.lastIsRunning, lastIsRunning == true {
```
Comment (joseph-elmallah) [inline Sources/UBQRScanner/QRScannerView.swift:53]: This is a good candidate for a guard structure

File: Sources/UBQRScanner/QRScannerView.swift [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+        let videoInput: AVCaptureDeviceInput
+        do {
+            videoInput = try AVCaptureDeviceInput(device: videoCaptureDevice)
+        } catch {
+            scanningDidFail(error: .captureSessionError(error))
+            return
+        }
+
+        if captureSession?.canAddInput(videoInput) ?? false {
```
Comment (joseph-elmallah) [inline Sources/UBQRScanner/QRScannerView.swift:161]: A guard is better suited here

File: Sources/UBQRScanner/QRScannerView.swift [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            captureSession?.addInput(videoInput)
+        } else {
+            scanningDidFail(error: .captureSessionError(nil))
+            return
+        }
+
+        let metadataOutput = AVCaptureMetadataOutput()
+
+        if captureSession?.canAddOutput(metadataOutput) ?? false {
```
Comment (joseph-elmallah) [inline Sources/UBQRScanner/QRScannerView.swift:170]: Guard

File: Sources/UBQRScanner/QRScannerView.swift [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    /// Whether the device has a torch that can be enabled
+    public var canEnableTorch: Bool {
+        guard let camera = videoCaptureDevice else { return false }
+        return camera.hasTorch && camera.isTorchAvailable
+    }
+
+    /// Start scanning, requesting the camera permission if needed
+    public func startScanning() {
```
Comment (joseph-elmallah) [inline Sources/UBQRScanner/QRScannerView.swift:90]: Have you tested calling startScanning before the view is added to the hierarchy? does this produce any failures? does it work after it is added without needing to call startScanning again?

File: Sources/UBQRScanner/QRScannerView.swift [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        setupCaptureSessionIfNeeded()
+
+        if let c = captureSession, !c.isRunning {
+            c.startRunning()
+        }
+    }
+
+    /// Turn the torch on or off
+    public func setTorch(on: Bool) {
+        guard let camera = videoCaptureDevice, canEnableTorch else { return }
```
Comment (joseph-elmallah) [inline Sources/UBQRScanner/QRScannerView.swift:101]: Can you enable torch without the camera be running? or before you call startScanning?

File: Sources/UBQRScanner/QRScannerView.swift (hunk 1 above)
Comment (joseph-elmallah) [inline Sources/UBQRScanner/QRScannerView.swift:17]: Do we want Obj-C support also?

File: Sources/UBQRScanner/QRScannerView.swift (hunk 5 above)
Comment (ubfelix) [reply Sources/UBQRScanner/QRScannerView.swift:90]: I just tested this and it doesn't seem to be a problem.

File: Sources/UBQRScanner/QRScannerView.swift (hunk 3 above)
Comment (ubfelix) [reply Sources/UBQRScanner/QRScannerView.swift]: Resolved.

File: Sources/UBQRScanner/QRScannerView.swift (hunk 4 above)
Comment (ubfelix) [reply Sources/UBQRScanner/QRScannerView.swift]: Resolved.

**Pull request merged**